### PR TITLE
Read, Write, Fsync, Setattr

### PR DESCRIPTION
**Pull request by mberhault**

### CODE REVIEW

File: filesystem/fs.go [hunk 1]
```diff
@@ -167,6 +167,84 @@ func (cfs CFS) list(parentID uint64) ([]fuse.Dirent, error) {
...
+
+	// Look for the original ID.
+	var id uint64
+	err = tx.QueryRow(lookupSql, oldParentID, oldName).Scan(&id)
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	// Delete it.
```
Comment (petermattis) [inline filesystem/fs.go]: If `oldParentID/oldName` exists you need to check it it is a directory and is empty.

File: filesystem/fs.go [hunk 2]
```diff
@@ -167,6 +167,84 @@ func (cfs CFS) list(parentID uint64) ([]fuse.Dirent, error) {
...
+INSERT INTO fs.namespace VALUES ($1, $2, $3);
+`
+		if _, err := tx.Exec(insertSql, newParentID, newName, id); err != nil {
+			_ = tx.Rollback()
+			return err
+		}
+
+	} else {
+		// Row found: delete the inode, and update the namespace entry.
+		const replaceSql = `
```
Comment (petermattis) [inline filesystem/fs.go]: Sorry, earlier comment was misplaced. If `newParentID/newName` exists you need to make sure it is an empty directory.

File: filesystem/fs.go [hunk 3]
```diff
@@ -167,6 +167,84 @@ func (cfs CFS) list(parentID uint64) ([]fuse.Dirent, error) {
...
+	if err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	// Delete it.
+	const deleteSql = `
+DELETE FROM fs.namespace WHERE (parentID, name) = ($1, $2);
+`
+	if _, err := tx.Exec(deleteSql, oldParentID, oldName); err != nil {
```
Comment (petermattis) [inline filesystem/fs.go]: Would be better to perform the `DELETE` in the same `Exec` as the `INSERT/UPDATE`. Reduce roundtrips.

File: filesystem/fs.go (hunk 2 above)
Comment (mberhault) [inline filesystem/fs.go]: it's a little more than that actually. If `oldParentID/oldName` is a directory, `newParentID/newName` must either not exist, or be an empty directory.
I need to refactor `fs.go` to have a wrapper transaction and a bunch of helper functions. It's getting rather ugly.

File: filesystem/fs.go (hunk 3 above)
Comment (mberhault) [inline filesystem/fs.go]: Done. Refactored `fs.go` to use some helpers and clean up things a bit, including two big sets of statements. one when the destination exists, the other when it does not.

File: filesystem/fs.go (hunk 2 above)
Comment (mberhault) [inline filesystem/fs.go]: Done.

File: filesystem/sql.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+// countChildren returns the number of children of inode with 'id'.
+func countChildren(e sqlExecutor, id uint64) (uint64, error) {
+	const countSql = `
+SELECT COUNT(parentID) FROM fs.namespace WHERE parentID = $1`
+
+	var count uint64
+	if err := e.QueryRow(countSql, id).Scan(&count); err != nil {
+		return 0, err
+	}
+	return count, nil
```
Comment (petermattis) [inline filesystem/sql.go]: You can return `count, err` in both cases:

```
  err := e.QueryRow(...).Scan(&count)
  return count, err
```

File: filesystem/sql.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+(SELECT id FROM fs.namespace WHERE (parentID, name) = ($1, $2))`
+	if err := e.QueryRow(sql, parentID, name).Scan(&raw); err != nil {
+		return nil, err
+	}
+
+	node := &Node{}
+	if err := json.Unmarshal([]byte(raw), node); err != nil {
+		return nil, err
+	}
+	return node, nil
```
Comment (petermattis) [inline filesystem/sql.go]: Similar to below, you can return `node, err` in both cases.

File: filesystem/node.go [hunk 6]
```diff
@@ -167,3 +206,68 @@ func (n Node) Remove(_ context.Context, req *fuse.RemoveRequest) error {
...
+
+	l := uint64(len(n.Data))
+	if newLen > l {
+		n.Data = append(n.Data, make([]byte, newLen-l)...)
+	}
+
+	written := copy(n.Data[req.Offset:], req.Data)
+	// TODO(marc): check that fuse forgets the node on errors, otherwise
+	// we'll have updated data in memory but not committed it.
+	if err := n.cfs.update(*n); err != nil {
```
Comment (petermattis) [inline filesystem/node.go]: You're storing all of the file data in the inode! I guess that works, but ugh. I was thinking of a setup where the file data was stored in another table that looks like:

```
CREATE TABLE filesystem.filedata (
  id       INT,
  block INT,
  data  BYTES,
  PRIMARY KEY(id, block)
)
```

We'd then set the block size to something like 4K or 16K.

**Comment by petermattis** [comment]
LGTM

File: filesystem/sql.go (hunk 4 above)
Comment (mberhault) [inline filesystem/sql.go]: Done. but with named returns.

File: filesystem/sql.go (hunk 5 above)
Comment (mberhault) [inline filesystem/sql.go]: Done.

File: filesystem/node.go (hunk 6 above)
Comment (mberhault) [inline filesystem/node.go]: I'll add a summary on main.go including a list of TODOs.

**Pull request merged**